From d325aa6e205ba7b9b681e774cf7c66ef8fc00800 Mon Sep 17 00:00:00 2001
From: Przemyslaw Czarnota <przemyslawx.czarnota@intel.com>
Date: Mon, 29 Jun 2020 14:17:20 +0200
Subject: [PATCH] PONRTSYS-7703: Ext VLAN rule: 0.7 - 'prio, dei and VLAN ID'
 fields are not contained after rule processed them

- Compare also the filtering part of the rule in dscp_merge()

The dscp_merge() was comparing only the treatment part of the Ext. Vlan
rule, resulting in an incorrect merging of the rules (IPv6 rules were merged
into IPv4). After this change the dscp_merge() will merge only rules
which have their filtering part equal.
---
 drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c | 23 ++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
index 2fa0d0298a31..214a0cdab7f7 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_ext_vlan.c
@@ -23,11 +23,26 @@
 
 static LIST_HEAD(tc_ext_vlan_storage);
 
-static int dscp_merge(struct dp_act_vlan *old,
+static int dp_pattern_vlan_eq(const struct dp_pattern_vlan *a,
+			      const struct dp_pattern_vlan *b)
+{
+	return a->prio == b->prio &&
+	       a->vid == b->vid &&
+	       a->tpid == b->tpid &&
+	       a->dei == b->dei &&
+	       a->proto == b->proto;
+}
+
+static int dscp_merge(struct dp_pattern_vlan *old_pattern,
+		      struct dp_act_vlan *old,
+		      struct dp_pattern_vlan *new_pattern,
 		      struct dp_act_vlan *new)
 {
 	int i;
 
+	if (!dp_pattern_vlan_eq(old_pattern, new_pattern))
+		return -1;
+
 	if (old->act != new->act)
 		return -1;
 
@@ -82,7 +97,7 @@ static int untagged_add(struct net_device *dev,
 
 	/* Attempt to merge new rule into existing one */
 	list_for_each_entry(p, &node->rules, list) {
-		if (!dscp_merge(&p->act, &rule->act)) {
+		if (!dscp_merge(&p->outer, &p->act, &rule->outer, &rule->act)) {
 			kfree(rule);
 			return 0;
 		}
@@ -123,7 +138,7 @@ static int single_tagged_add(struct net_device *dev,
 
 	/* Attempt to merge new rule into existing one */
 	list_for_each_entry(p, &node->rules, list) {
-		if (!dscp_merge(&p->act, &rule->act)) {
+		if (!dscp_merge(&p->outer, &p->act, &rule->outer, &rule->act)) {
 			kfree(rule);
 			return 0;
 		}
@@ -164,7 +179,7 @@ static int double_tagged_add(struct net_device *dev,
 
 	/* Attempt to merge new rule into existing one */
 	list_for_each_entry(p, &node->rules, list) {
-		if (!dscp_merge(&p->act, &rule->act)) {
+		if (!dscp_merge(&p->outer, &p->act, &rule->outer, &rule->act)) {
 			kfree(rule);
 			return 0;
 		}
